pymtl3/datatypes/bit_structs.py

```diff
@@ -0,0 +1,476 @@
+# following:
+#
+# def __eq__( self, other ):
+#   if other.__class__ is self.__class__:
+#     return (self.x,self.y,) == (other.x,other.y,)
+#   else:
+#     raise NotImplemented
+
+def _mk_eq_fn( fields ):
+  self_tuple  = _mk_tuple_str( 'self', fields )
+  other_tuple = _mk_tuple_str( 'other', fields )
+  return _create_fn(
+    '__eq__',
+    [ 'self', 'other' ],
+    [ 'if other.__class__ is self.__class__:',
+      f'  return {self_tuple} == {other_tuple}',
+      'else:',
+      '  raise NotImplemented']
+  )
+
+#-------------------------------------------------------------------------
+# _mk_hash_fn
+#-------------------------------------------------------------------------
+# Creates a __hash__ function based on fields. By default it just hashes
+# all fields. For example, if fields contains two field x (Bits4) and y
+# (Bits4), _mk_init_fn will return a function that looks like the
+# following:
+#
+# def __hash__( self ):
+#   return hash((self.x,self.y,))
+
+def _mk_hash_fn( fields ):
+    self_tuple = _mk_tuple_str( 'self', fields )
+    return _create_fn(
+      '__hash__',
+      [ 'self' ],
+      [ f'return hash({self_tuple})' ]
+    )
+
+#-------------------------------------------------------------------------
+# _get_field
+#-------------------------------------------------------------------------
+# Extract field information from cls based on annotations.
+
+def _get_field( cls, a_name, a_type ):
+  default = getattr( cls, a_name, MISSING )
+  f = field( a_type, default=default )
+  f.name = a_name
+  # Yanghui: Should we check default must be instances of Bits or bit
+  # struct here?
+  return f
+
+#-------------------------------------------------------------------------
+# _get_self_name
+#-------------------------------------------------------------------------
+# Return a self name based on fields.
+
+def _get_self_name( fields ):
+  return( _ANTI_CONFLICT_SELF_NAME if _DEFAULT_SELF_NAME in fields else
+          _DEFAULT_SELF_NAME )
+
+#-------------------------------------------------------------------------
+# _process_cls
+#-------------------------------------------------------------------------
+# Process the input cls and add methods to it.
+
+def _process_class( cls, add_init=True, add_str=True, add_repr=True,
+                    add_hash=True ):
+
+  fields = {}
+  add_eq = True
+
+  # Get annotations of the class
+  cls_annotations = cls.__dict__.get('__annotations__', {})
+  if not cls_annotations:
+    raise NoFieldDeclaredError(
+      'No field is delacred in the bit struct definition' )
+
+  # Get field information from the annotation
+  cls_fields = [ _get_field( cls, a_name, a_type )
+                 for a_name, a_type in cls_annotations.items() ]
+
+  # Create a dictionary of fields
+  for f in cls_fields:
+    fields[ f.name ] = f
+
+  # Stamp the special attribute so that translation pass can identify it
+  # as bit struct.
+  setattr( cls, _IS_BIT_STRUCT, fields )
```

Comment:
No need to use setattr here because the attribute is a constant string

Reply:
I keep it this way so that we can easily change this constant string. This won't hurt performance right?

Reply:
Since it's not during runtime, it shouldn't hurt.